File: src/core/common/clock.cc
#include "core/common/clock.h"

#include <atomic>
#include <windows.h>

namespace wincap {

namespace {
std::atomic<std::uint64_t> g_freq{0};

std::uint64_t LoadFreq() noexcept {
    std::uint64_t f = g_freq.load(std::memory_order_acquire);
    if (f != 0) return f;
    LARGE_INTEGER li{};
    QueryPerformanceFrequency(&li);
    f = static_cast<std::uint64_t>(li.QuadPart);
    g_freq.store(f, std::memory_order_release);
    return f;
}
} // namespace

void Clock::Init() noexcept { (void)LoadFreq(); }

std::uint64_t Clock::Frequency() noexcept { return LoadFreq(); }

std::uint64_t Clock::NowTicks() noexcept {
    LARGE_INTEGER li{};
    QueryPerformanceCounter(&li);
    return static_cast<std::uint64_t>(li.QuadPart);
}

std::uint64_t Clock::TicksToNs(std::uint64_t ticks) noexcept {
    const std::uint64_t freq = LoadFreq();
    // Split to avoid 64-bit overflow on large counters:
    //   ns = (whole_seconds * 1e9) + (remainder * 1e9 / freq)
    const std::uint64_t whole = ticks / freq;
    const std::uint64_t rem   = ticks % freq;
    return whole * 1'000'000'000ull + (rem * 1'000'000'000ull) / freq;
}

std::uint64_t Clock::NowNs() noexcept { return TicksToNs(NowTicks()); }

} // namespace wincap

File: crates/wincap-core/src/clock.rs
use std::sync::OnceLock;
use windows::Win32::System::Performance::{QueryPerformanceCounter, QueryPerformanceFrequency};

static FREQUENCY: OnceLock<u64> = OnceLock::new();

fn frequency() -> u64 {
    *FREQUENCY.get_or_init(|| {
        let mut freq = 0i64;
        unsafe { QueryPerformanceFrequency(&mut freq).unwrap() };
        freq as u64
    })
}

/// Initialise the global clock. Idempotent, thread-safe.
pub fn init() {
    let _ = frequency();
}

/// Current QPC counter value.
pub fn now_ticks() -> u64 {
    let mut ticks = 0i64;
    unsafe { QueryPerformanceCounter(&mut ticks).unwrap() };
    ticks as u64
}

/// Convert raw QPC ticks to nanoseconds. Splits to avoid 64-bit overflow.
pub fn ticks_to_ns(ticks: u64) -> u64 {
    let freq = frequency();
    let whole = ticks / freq;
    let rem = ticks % freq;
    whole * 1_000_000_000 + (rem * 1_000_000_000) / freq
}

/// Current time in nanoseconds since the QPC epoch.
pub fn now_ns() -> u64 {
    ticks_to_ns(now_ticks())
}

/// Convert a WinRT TimeSpan (100-ns units) to nanoseconds.
pub fn hundred_ns_to_ns(hns: i64) -> u64 {
    (hns as u64) * 100
}
